src/shogun/io/streaming/ParseBuffer.h

```diff
@@ -99,11 +99,14 @@ template <class T> class CParseBuffer: public CSGObject
 	Example<T>* get_free_example()
 	{
 #ifdef HAVE_CXX11
-		std::unique_lock<std::mutex> write_lk(*write_mutex, std::defer_lock);
-		std::unique_lock<std::mutex> current_ex_lock(*ex_in_use_mutex[ex_write_index], std::defer_lock);
-		std::lock(write_lk, current_ex_lock);
-		while (ex_used[ex_write_index] == E_NOT_USED)
-			ex_in_use_cond[ex_write_index]->wait(current_ex_lock);
+//		std::unique_lock<std::mutex> write_lk(*write_mutex, std::defer_lock);
+//		std::unique_lock<std::mutex> current_ex_lock(*ex_in_use_mutex[ex_write_index], std::defer_lock);
+//		std::lock(write_lk, current_ex_lock);
+        int32_t current_write_index = ex_write_index.load(std::memory_order_relaxed);  // only written from parser thread
+        int32_t current_read_index = ex_read_index.load(std::memory_order_acquire); // for synchronisation
+		while (ex_used[current_write_index] == E_NOT_USED)
```

Comment:
i reckon there should be a more elegant way to do this... this is kind of a bruteforce solution. btw have you profiled it, whether this actually results in a speedup (or maybe a slowdown?)
definitely fences should be used since you dont use the values that you load. although i'm thinking whether somehow using exchange could be used here....